src/main/java/de/uos/informatik/ko/rcp/geneticAlgo/GeneratePop.java

```diff
@@ -0,0 +1,101 @@
+package de.uos.informatik.ko.rcp.geneticAlgo;
+
+import java.lang.reflect.Array;
+import java.lang.Integer;
+import java.util.*;
+import de.uos.informatik.ko.rcp.Utils;
+import de.uos.informatik.ko.rcp.Instance;
+
+
+public class GeneratePop {
+
+    private GeneratePop() {}
+    //change to Hashset of ArrayLists
+    public static HashSet<ArrayList<Integer>> generatePop(Instance instance, Integer pop_size,Random myGenerator){
+        // all activities without predecessor are possible candidates
+        HashSet<ArrayList<Integer>> results = new HashSet<ArrayList<Integer>>();
+        HashMap<Integer,HashSet<Integer>> mypredecessors;
+        HashSet<Integer> candidates = new HashSet<Integer>();
+        //scheduled activities
+        ArrayList<Integer> scheduled = new ArrayList<Integer>();
+        for (int i=0; i< pop_size*5; i++) {
+            mypredecessors = Utils.buildPredecessorMap(instance);
+            for (Map.Entry<Integer, HashSet<Integer>> entry : mypredecessors.entrySet()) {
+                if (entry.getValue().isEmpty() && (!(scheduled.contains(entry.getKey())))) {
+                    candidates.add(entry.getKey());
+                }
+            }
+            Integer mycandidate = chooseone(candidates, myGenerator);
+            scheduled.add(mycandidate);
+            candidates.remove(mycandidate);
+            //delete scheduled activity from predecessor list
+            for (Map.Entry<Integer, HashSet<Integer>> entry : mypredecessors.entrySet()) {
+                entry.getValue().remove(mycandidate);
+            }
+            while (scheduled.size() < instance.n()) {
+                for (Map.Entry<Integer, HashSet<Integer>> entry : mypredecessors.entrySet()) {
+                    if (entry.getValue().isEmpty()&& (!(scheduled.contains(entry.getKey())))) {
+                        candidates.add(entry.getKey());
+                    }
+                }
+                Integer actcandidate = chooseone(candidates, myGenerator);
+                scheduled.add(actcandidate);
+                candidates.remove(actcandidate);
+                // delete scheduled activity from predecessor list
+                for (Map.Entry<Integer, HashSet<Integer>> entry : mypredecessors.entrySet()) {
+                    entry.getValue().remove(actcandidate);
+                }
+            }
+            ArrayList<Integer> resultsList = new ArrayList<Integer>();
+            for (Integer elem: scheduled){
+                resultsList.add(elem);
+            }
+            results.add(resultsList);
+            scheduled.clear();
+            candidates.clear();
+            if (results.size() >= pop_size){
+                break;
+            }
```

Comment:
Why is this needed? Shouldn't the loop condition take care of this already?

Reply:
In the current version not. I think there are prettier ways to implement the outer loop. Currently the for loop just consists of pop_size*(a_new_variable_not_five) iterations. If the algorithm generates pop_size unique elements, it should stop iterating, since it already created enough elements.

Reply:
Yep, https://github.com/leon-richardt/rcpsp-solver/pull/5#discussion_r637578921 cleared this up for me :+1: Maybe a comment could be added somewhere explaining that member generation will be attempted again for non-unique activity lists